### PR TITLE
Google Sitemap refactoring - introduce item resolver

### PR DESCRIPTION
**Pull request by piotrkwiecinski**
Refactoring to introduce injectable item resolver as way to add items to sitemap

### Description
- Introduce SitemapItemResolverInterface with concrete implementations for cms pages, products, categories and aggregate
- Introduce SitemapItemInterface as representation of single sitemap item
- Introduce SitemapConfigReaderInterface
- Introduce SitemapItemResolverConfigInterface
- Remove addSitemapItems as it was introduced in develop branch and wasn't used yet as public API
- Deprecate helper (suggested by @vrann). Getting data from configuration is part of introduced concrete implementations.

### Fixed Issues (if relevant)

1. magento/magento2#10045: Refactoring: Adding items to the Sitemap

### Manual testing scenarios

1. Go to Marketing > SEO > Sitemap
2. Add sitemap - provide path and filename
3. Click generate & save sitemap
4. Check if content is properly generated.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Sitemap/Model/ProductSitemapItemResolver.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        $this->itemFactory = $itemFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems($storeId)
+    {
+        $collection = $this->productFactory->create()->getCollection($storeId);
+        $items = array_map(function ($item) use ($storeId) {
```
Comment (Koc) [inline app/code/Magento/Sitemap/Model/ProductSitemapItemResolver.php]: Does we need exclude disabled products also? And this code will not work on large collections

**Comment by okorshenko** [comment]
Hi @piotrkwiecinski Could you please look at broken tests?

**Comment by piotrkwiecinski** [comment]
@vrann, @okorshenko tests should be green now. Last one was static.

**Comment by vrann** [comment]
@piotrkwiecinski finished code review, looks good. Running internal builds.

**Review by vrann — APPROVED** [review]

File: app/code/Magento/Sitemap/Model/Sitemap.php [hunk 2]
```diff
@@ -218,81 +241,32 @@ public function __construct(
...
         if ($this->_stream) {
             return $this->_stream;
         } else {
-            throw new \Magento\Framework\Exception\LocalizedException(__('File handler unreachable'));
+            throw new LocalizedException(__('File handler unreachable'));
         }
     }
 
     /**
-     * Add a sitemap item to the array of sitemap items
```
Comment (vrann) [inline app/code/Magento/Sitemap/Model/Sitemap.php]: Please restore these two public methods:
- they are in the class marked as @api
- the class will be released as is in 2.2 release
- current pull request will get to the 2.3 release

**Review by vrann — CHANGES_REQUESTED** [review]

**Comment by Ctucker9233** [comment]
@magento-team Will this get backported to 2.2?

**Comment by piotrkwiecinski** [comment]
If it would be useful and there is no objections from  @magento-team I could port it back on Monday or Tuesday

**Comment by Ctucker9233** [comment]
@piotrekkaminski Thanks

**Pull request merged**